Renderer qt

jmercouris opened the pull request:
Incomplete,

questions:

how shall we organize the ASDF file? For example, do we have a next/gtk and a next/qt system? Do we have some sort of feature tag?

Ambrevar commented:
What we want to offer, if I got your question right, is build-time conditionals on the renderers the include.

I'm no ASDF expert, so maybe we could ask the devs.  Or maybe Lem's .asd has something similar.

- With multiple systems, we would have one system per renderer + one system for the rest of the source, + 1 system for all combinations (e.g. gtk+qt, gi-gtk+qt, etc.).  The combinatorial definition of systems is a bit unfortunate.

- Have one system with with dynamic components.  We could choose the components according to, say, an environment variable.
For instance, if NEXT_RENDERER_GTK and NEXT_RENDERER_GI_QT are set, include those two.  Should be simple enough and composes well.

jmercouris commented:
> What we want to offer, if I got your question right, is build-time conditionals on the renderers the include.

Exactly what I am wondering about

> I'm no ASDF expert, so maybe we could ask the devs. Or maybe Lem's .asd has something similar.
> 
> * With multiple systems, we would have one system per renderer + one system for the rest of the source, + 1 system for all combinations (e.g. gtk+qt, gi-gtk+qt, etc.).  The combinatorial definition of systems is a bit unfortunate.
> * Have one system with with dynamic components.  We could choose the components according to, say, an environment variable.
>   For instance, if NEXT_RENDERER_GTK and NEXT_RENDERER_GI_QT are set, include those two.  Should be simple enough and composes well.

That would be nice, I also don't remember how to do it, though I believe there is a way.

jmercouris commented:
I looked at old versions of the asd, there is if-feature, but it will not suffice for our needs. I don't believe we can set the values via an environment variable or anything. A system for each derviative seems to be the best.

jmercouris commented:
@Ambrevar 

It is now ready for a first cursory glance. Please took a look at the prototype methods in `browser.lisp` and if you are curious, the implementation details in `cl-webengine`. Also, if you could please fix some of the conflicts that I've had to resolve in a more elegant way (for example, removing the export of the class gtk-browser, because it causes conflicts (on darwin)).

As soon as the keys are captured, the first alpha will be ready! :-)

jmercouris commented:
P.S. Feel free to push fixes directly onto this branch!

Ambrevar commented:
I believe we can do a lot with ASDF, including running (uiop:getenv).
I've never done it though.
We could have a look at SLIME or SLY .asd files, they are rather sophisticated.

jmercouris commented:
Yes, but I received a lot of kickback about making the ASD files complex on `#lisp`... :-D

makes them difficult to analyze

Ambrevar commented:
We don't have to make them complex.

jmercouris commented:
I think the separate systems is very neat anyways. Take a look at it, I prefer it.

jmercouris commented:
Yay! Keyboard input works! With many exceptions of course... but it works!

Ambrevar commented:
Fantastic, I'll test as soon as I find the time!

jmercouris commented:
I'm now trying to update the developer documentation, and trying to get most keyboard input working correctly. There are still some strange characters being emitted, and then I would say we are at an alpha! I would suggest that we make the WebKit version the default still in the makefile since it is a much better port.

jmercouris commented:
I think this is ready to be merged onto master! Yes, it is not as complete as the GTK version, but it is functional :-)

jmercouris commented:
And, importantly, it does not break the GTK version :-D

Ambrevar commented:
Wait, I've spotted a few issues, I'll reports in a little while.

jmercouris commented:
You can also just fix them, unless they are not clear decisions

Most of the work needs to be done now in cl-webengine to fix some odd behaviors with keybindings particularly. Also some bugs in Qt itself...

Ambrevar commented:
MANUAL.org: You've removed the StumpWM section, however it is still valid: this is a general problem with GTK programs, it is not related to our interception of mouse events.

You've also remove the local properties I had added here, which nicely auto-update the TOC upon saving.  I think it's a nice feature and I'd like to leave it on.  Any counter points?

jmercouris commented on source/start.lisp:
```diff
@@ -229,8 +229,8 @@ EXPR must contain one single Lisp form. Use `progn' if needed."
-                       (gtk:within-gtk-thread
-                         (open-urls urls)))
+                       ;; TODO: execute within main thread to avoid crashing on Linux
+                       (open-urls urls))
```
That is a good idea, what if the user is running both renderers at the same time?

jmercouris commented:
> MANUAL.org: You've removed the StumpWM section, however it is still valid: this is a general problem with GTK programs, it is not related to our interception of mouse events.

Oh, I did not know! In any case, is that not a problem with StumpWM rather than Next?
 
> You've also remove the local properties I had added here, which nicely auto-update the TOC upon saving. I think it's a nice feature and I'd like to leave it on. Any counter points?

That was an accident, I didn't even know I did that! I bet that my org-mode deleted them or something, can you please restore them, as I don't know what they were/where they were?

Ambrevar commented:
> Oh, I did not know! In any case, is that not a problem with StumpWM rather than Next?

Yes, but we've been asked about it before, so a quick troubleshooting on
our side
doesn't hurt.

> That was an accident, I didn't even know I did that! I bet that my org-mode deleted them or something, can you please restore them, as I don't know what they were/where they were?

Will do!

Ambrevar commented:
Done:
- Fixed the doc.
- Fixed the warnings.
- Added macro to define ffi methods concisely.

What's missing: For now we need to build GTK by default and enable it.
How do you propose we do that?
We can make the :next system a dependency of :next/gtk and build the
binary in next/gtk instead.

jmercouris commented:
> Done: - Fixed the doc. - Fixed the warnings. - Added macro to define ffi methods concisely. What's missing: For now we need to build GTK by default and enable it. How do you propose we do that? We can make the :next system a dependency of :next/gtk and build the binary in next/gtk instead.

Hello Pierre,

thank you for the macro! I was thinking myself that it was a good opportunity for a macro.

as per your question, I don't understand. `next/gtk` already depends on Next, definition:

```
(asdf:defsystem :next/gtk
  :depends-on (:next
               :cl-cffi-gtk
               :cl-webkit2)
  :pathname "source"
  :components ((:file "renderer-gtk")))
```

You can just do `(ql:quickload :next/gtk)` you don't have to first load `:next` and then `:gtk`. All we would have to do is adjust the makefile to build `next/gtk` by default, as well as possibly move the `:build-pathname`, `:entry-point` and `build-operation`.

Thoughts?

jmercouris commented on source/renderer-qt.lisp:
```diff
@@ -153,3 +153,8 @@
+
+(defmethod ffi-within-renderer-thread ((browser qt-browser) thunk)
+  (declare (ignore browser))
+  ;; TODO: Test this!
```
This would probably have to be `trivial-main-thread:within-main-thread`. Qt itself considers the thread it starts on to be the "main" thread. So we should probably start it on the main thread, and add operations to it on the main thread, or otherwise find a way to keep track of the thread it is on. Please note that on macOS it *must* start on the literal main thread.

Ambrevar commented:
> You can just do `(ql:quickload :next/gtk)` you don't have to first
> load `:next` and then `:gtk`. All we would have to do is adjust the
> makefile to build `next/gtk` by default, as well as possibly move the
> `:build-pathname`, `:entry-point` and `build-operation`.

Absolutely.

jmercouris commented:
Done! I will be merging this shortly, there is one more thing I would like to try...
